services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -0,0 +1,1109 @@
+            path: self.path,
+        }))
+    }
+
+    fn type_name(&self) -> String {
+        "NetCdfCfProviderDefinition".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+pub struct NetCdfCfDataProvider {
+    id: DatasetProviderId,
+    path: PathBuf,
+}
+
+/// TODO: This should be part of the GDAL crate
+/// TODO: free group with `GDALGroupRelease()`
+#[derive(Debug)]
+struct MdGroup<'d> {
+    c_group: *mut GDALGroupHS,
+    _dataset: &'d Dataset,
+    name: String,
+}
+
+/// TODO: This should be part of the GDAL crate
+/// TODO: free group with `GDALGroupRelease()`
+#[derive(Debug)]
+struct MdArray<'g> {
+    c_mdarray: GDALMDArrayH,
+    _group: &'g MdGroup<'g>,
+}
+
+/// TODO: There are potential memory leaks from the C-API
+impl<'d> MdGroup<'d> {
+    unsafe fn _string(raw_ptr: *const c_char) -> String {
+        let c_str = CStr::from_ptr(raw_ptr);
+        c_str.to_string_lossy().into_owned()
+    }
+
+    unsafe fn _string_array(raw_ptr: *mut *mut c_char) -> Vec<String> {
+        let mut ret_val: Vec<String> = vec![];
+        let mut i = 0;
+        loop {
+            let ptr = raw_ptr.add(i);
+            if ptr.is_null() {
+                break;
+            }
+            let next = ptr.read();
+            if next.is_null() {
+                break;
+            }
+            let value = Self::_string(next);
+            i += 1;
+            ret_val.push(value);
+        }
+        ret_val
+    }
+
+    unsafe fn _last_null_pointer_err(method_name: &'static str) -> GdalError {
+        let last_err_msg = Self::_string(gdal_sys::CPLGetLastErrorMsg());
+        gdal_sys::CPLErrorReset();
+        GdalError::NullPointer {
+            method_name,
+            msg: last_err_msg,
+        }
+    }
+
+    unsafe fn _last_cpl_err(cpl_err_class: CPLErr::Type) -> GdalError {
+        let last_err_no = gdal_sys::CPLGetLastErrorNo();
+        let last_err_msg = Self::_string(gdal_sys::CPLGetLastErrorMsg());
+        gdal_sys::CPLErrorReset();
+        GdalError::CplError {
+            class: cpl_err_class,
+            number: last_err_no,
+            msg: last_err_msg,
+        }
+    }
+
+    fn from_dataset(dataset: &'d Dataset) -> Result<Self> {
+        let c_group = unsafe {
+            let c_group = GDALDatasetGetRootGroup(dataset.c_dataset());
+
+            if c_group.is_null() {
+                return Err(NetCdfCf4DProviderError::GdalMd {
+                    source: Self::_last_null_pointer_err("GDALGetRasterBand"),
+                });
+            }
+
+            c_group
+        };
+
+        Ok(Self {
+            c_group,
+            _dataset: dataset,
+            name: "".to_owned(),
+        })
+    }
+
+    /// Don't put a NULL-byte in the name!!!
+    fn attribute_as_string(&self, name: &str) -> Result<String, GdalError> {
+        let name = CString::new(name).expect("no null-byte in name");
```

Comment:
you should check that name is not empty

Reply:
Doesn't GDAL just return an error if it is not possible to open it?

Reply:
i don't know. you should  check that.

Reply:
https://gdal.org/api/gdalattribute_cpp.html#_CPPv4NK17GDALIHasAttribute13GetAttributesE12CSLConstList

> Returns    the new attribute, or nullptr if case of error

And we check for `is_null` and return an error.